File: src/checker/sources/brew-api.ts
import type { VersionSource, VersionSourceResult, VersionInfo } from './index.js';
import type { BrewSourceConfig } from '../../config.js';
import { fetchWithEtag } from './base.js';

export type { BrewSourceConfig } from '../../config.js';

export class BrewSource implements VersionSource {
  readonly name = 'brew';
  private readonly config: BrewSourceConfig;

  constructor(config: BrewSourceConfig) {
    this.config = config;
  }

  async fetchLatest(options?: {
    etag?: string;
    signal?: AbortSignal;
  }): Promise<VersionSourceResult> {
    const url = `https://formulae.brew.sh/api/cask/${this.config.caskName}.json`;

    const result = await fetchWithEtag({ url }, options);

    if (result.kind !== 'response') {
      if (result.kind === 'error' && result.status === 404) {
        return { kind: 'error', reason: `Homebrew cask not found: ${this.config.caskName}`, status: 404 };
      }
      return result;
    }

    const data = await result.response.json();

    if (typeof data.version !== 'string' || !data.version) {
      return {
        kind: 'error',
        reason: 'Homebrew API response missing version field',
      };
    }

    const info: VersionInfo = {
      version: data.version,
      releaseUrl: data.homepage,
    };

    return { kind: 'found', info, etag: result.etag };
  }
}

File: rust/crates/update-kit/src/checker/sources/brew_api.rs
use super::{FetchOptions, VersionInfo, VersionSource, VersionSourceResult};

/// A version source backed by Homebrew (brew info --json).
pub struct BrewSource {
    cask_name: String,
}

impl BrewSource {
    pub fn new(cask_name: String) -> Self {
        Self { cask_name }
    }
}

#[async_trait::async_trait]
impl VersionSource for BrewSource {
    fn name(&self) -> &str {
        "brew"
    }

    async fn fetch_latest(&self, _options: FetchOptions) -> VersionSourceResult {
        let output = match std::process::Command::new("brew")
            .args(["info", "--json=v2", "--cask", &self.cask_name])
            .output()
        {
            Ok(o) => o,
            Err(e) => {
                return VersionSourceResult::Error {
                    reason: format!("Failed to run brew: {}", e),
                    status: None,
                }
            }
        };

        if !output.status.success() {
            let stderr = String::from_utf8_lossy(&output.stderr);
            return VersionSourceResult::Error {
                reason: format!("brew info failed: {}", stderr.trim()),
                status: None,
            };
        }

        let json: serde_json::Value = match serde_json::from_slice(&output.stdout) {
            Ok(j) => j,
            Err(e) => {
                return VersionSourceResult::Error {
                    reason: format!("Failed to parse brew JSON: {}", e),
                    status: None,
                }
            }
        };

        // brew info --json=v2 returns { "casks": [ { "version": "...", ... } ] }
        let version = json
            .get("casks")
            .and_then(|v| v.as_array())
            .and_then(|arr| arr.first())
            .and_then(|cask| cask.get("version"))
            .and_then(|v| v.as_str())
            .map(String::from);

        match version {
            Some(version) => VersionSourceResult::Found {
                info: VersionInfo {
                    version,
                    release_url: None,
                    release_notes: None,
                    assets: None,
                    published_at: None,
                },
                etag: None,
            },
            None => VersionSourceResult::Error {
                reason: "Could not find cask version in brew output".into(),
                status: None,
            },
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_source_name() {
        let source = BrewSource::new("my-cask".into());
        assert_eq!(source.name(), "brew");
    }
}
